File: simulators/de.unistuttgart.iste.sqa.mpw.hamstersimulator.java/core/src/test/java/de/unistuttgart/hamster/view/HamsterViewTestBase.java
package de.unistuttgart.hamster.view;

import de.unistuttgart.hamster.facade.TerritoryLoader;
import de.unistuttgart.hamster.facade.Hamster;
import de.unistuttgart.hamster.facade.HamsterGame;
import de.unistuttgart.hamster.viewmodel.impl.HamsterGameViewPresenter;
import de.unistuttgart.iste.sqa.mpw.framework.viewmodel.GameViewInput;
import de.unistuttgart.iste.sqa.mpw.framework.viewmodel.GameViewModel;
import de.unistuttgart.iste.sqa.mpw.framework.viewmodel.ViewModelCell;
import de.unistuttgart.iste.sqa.mpw.framework.viewmodel.ViewModelRow;
import org.junit.jupiter.api.BeforeAll;

import java.io.IOException;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import static org.junit.jupiter.api.Assertions.assertEquals;

public class HamsterViewTestBase {

    private int maxCharsPerCell = 4;
    private static Map<String, String> characterMap = new HashMap<>();
    private HamsterGame game;
    protected Hamster paule;
    protected GameViewInput viewInput;
    protected GameViewModel viewModel;

    @BeforeAll
    public static void setup() {
        initCharMapping();
    }

    protected void withTerritory(final String fileName) throws IOException {
        final String path = getResourcePath(fileName);
        game = new HamsterGame();
        TerritoryLoader.initializeFor(game).loadFromResourceFile(path);

        game.startGame();
        game.getPerformance().setDelayEnabled(false);

        var territory = game.getTerritory();
        paule = territory.getDefaultHamster();

        var presenter = new HamsterGameViewPresenter(game);
        presenter.bind();
        viewInput = presenter;
        viewModel = presenter.getViewModel();

        game.setUserInputInterface(new UserInputInterfaceFake());
    }

    protected void initializeOtherTerritory(final String fileName) throws IOException {
        game.hardReset();
        TerritoryLoader.initializeFor(game).loadFromResourceFile(getResourcePath(fileName));
    }

    private String getResourcePath(String fileName) {
        return "de.unistuttgart.hamster.territories/" + fileName;
    }

    protected void clickPlay() {
        viewInput.playClicked();
    }

    protected void clickPause() {
        viewInput.pauseClicked();
    }

    protected void clickUndo() {
        viewInput.undoClicked();
    }

    protected void clickRedo() {
        viewInput.redoClicked();
    }

    protected void assertTerritory(String expected) {
        var actual = new HamsterViewModelStringifier(characterMap, maxCharsPerCell)
                .territoryToExpectationString(viewModel);
        assertEquals(expected, actual);
        assertLocationsAreSet();
        assertSizeIsConsistent();
    }

    protected void assertButtons(String expected) {
        var actual = HamsterViewModelStringifier.buttonBarToExpectationString(viewModel);
        assertEquals(expected, actual);
        assertLocationsAreSet();
        assertSizeIsConsistent();
    }

    private void assertLocationsAreSet() {
        List<ViewModelRow> rows = viewModel.getRows();
        for (int rowIndex = 0; rowIndex < rows.size(); rowIndex++) {
            ViewModelRow row = rows.get(rowIndex);
            List<ViewModelCell> cells = row.getCells();
            for (int columnIndex = 0; columnIndex < cells.size(); columnIndex++) {
                ViewModelCell cell = cells.get(columnIndex);
                assertEquals(columnIndex, cell.getLocation().getColumn());
                assertEquals(rowIndex, cell.getLocation().getRow());
            }

        }

    }

    private void assertSizeIsConsistent() {
        List<ViewModelRow> rows = viewModel.getRows();
        assertEquals(rows.size(), viewModel.getSize().getRowCount());
        for (ViewModelRow row : rows) {
            assertEquals(row.getCells().size(), viewModel.getSize().getColumnCount());
        }
    }

    protected void assertLog(String expected) {
        String actual = HamsterViewModelStringifier.logToString(viewModel);
        assertEquals(expected, actual);
    }

    private static void initCharMapping() {
        characterMap.put("1Corn", " 1*");
        characterMap.put("2Corn", " 2*");
        characterMap.put("3Corn", " 3*");
        characterMap.put("4Corn", " 4*");
        characterMap.put("5Corn", " 5*");
        characterMap.put("6Corn", " 6*");
        characterMap.put("7Corn", " 7*");
        characterMap.put("8Corn", " 8*");
        characterMap.put("9Corn", " 9*");
        characterMap.put("10Corn", "10*");
        characterMap.put("11Corn", "11*");
        characterMap.put("12Corn", "12*");
        characterMap.put("12PlusCorn", "12+");
        characterMap.put("Hamster", ">");
        characterMap.put("Hamster[90]", "v");
        characterMap.put("Hamster[180]", "<");
        characterMap.put("Hamster[270]", "^");
        characterMap.put("Wall", "####");
    }

}

File: core/test/view/HamsterViewTestBase.cpp
#include "HamsterViewTestBase.h"

#include "HamsterGameViewPresenter.h"
#include "UserInputInterfaceFake.h"
#include "HamsterViewModelStringifier.h"

#include <memory>

using namespace hamster;
using namespace viewmodel;

std::unordered_map<std::string, std::string> HamsterViewTestBase::characterMap;

void HamsterViewTestBase::SetUp() {
    initCharMapping();
}

void HamsterViewTestBase::TearDown() {
    game->getPerformance()->abortOrStopGame();
}

void HamsterViewTestBase::withTerritory(const std::string& path) {
    game = HamsterGame::create();
    TerritoryLoader::initializeFor(*game)->loadFromResourceFile("resources" + path);

    game->getPerformance()->setDelayEnabled(false);
    game->startGame();

    auto territory = game->getTerritory();
    paule = territory->getDefaultHamster();

    presenter = std::make_shared<HamsterGameViewPresenter>(game);
    presenter->bind();
    viewInput = presenter;
    viewModel = presenter->getViewModel();

    game->setUserInputInterface(std::make_shared<UserInputInterfaceFake>());
}

void HamsterViewTestBase::initializeOtherTerritory(const std::string& path) {
    game->hardReset();
    TerritoryLoader::initializeFor(*game)->loadFromResourceFile("resources" + path);
}

void HamsterViewTestBase::clickPlay() {
    viewInput->playClicked();
}

void HamsterViewTestBase::clickPause() {
    viewInput->pauseClicked();
}

void HamsterViewTestBase::clickUndo() {
    viewInput->undoClicked();
}

void HamsterViewTestBase::clickRedo() {
    viewInput->redoClicked();
}

void HamsterViewTestBase::assertTerritory(const std::string& expected) {
    std::string actual = HamsterViewModelStringifier(characterMap, maxCharsPerCell).territoryToExpectationString(*viewModel);
    EXPECT_EQ(expected, actual);
    assertLocationsAreSet();
    assertSizeIsConsistent();
}

void HamsterViewTestBase::assertLocationsAreSet() {
    int rowIndex = 0;
    for (auto& row : viewModel->getRows()) {
        int columnIndex = 0;
        for (auto& cell : row.getCells()) {
            EXPECT_EQ(columnIndex, cell.getLocation().getColumn());
            EXPECT_EQ(rowIndex, cell.getLocation().getRow());
            columnIndex++;
        }
        rowIndex++;
    }
}

void HamsterViewTestBase::assertSizeIsConsistent() {
    auto& rows = viewModel->getRows();
    EXPECT_EQ(rows.size(), viewModel->getSize().getRowCount());
    for (auto& row : rows) {
        EXPECT_EQ(row.getCells().size(), viewModel->getSize().getColumnCount());
    }
}

void HamsterViewTestBase::assertLog(const std::string& expected) {
    std::string actual = HamsterViewModelStringifier::logToString(*viewModel);
    EXPECT_EQ(expected, actual);
}

void HamsterViewTestBase::initCharMapping() {
    characterMap["1Corn32"] = " 1*";
    characterMap["2Corn32"] = " 2*";
    characterMap["3Corn32"] = " 3*";
    characterMap["4Corn32"] = " 4*";
    characterMap["5Corn32"] = " 5*";
    characterMap["6Corn32"] = " 6*";
    characterMap["7Corn32"] = " 7*";
    characterMap["8Corn32"] = " 8*";
    characterMap["9Corn32"] = " 9*";
    characterMap["10Corn32"] = "10*";
    characterMap["11Corn32"] = "11*";
    characterMap["12Corn32"] = "12*";
    characterMap["12PlusCorn32"] = "12+";
    characterMap["Hamster32"] = ">";
    characterMap["Hamster32[90]"] = "v";
    characterMap["Hamster32[180]"] = "<";
    characterMap["Hamster32[270]"] = "^";
    characterMap["Wall32"] = "####";
}

void HamsterViewTestBase::createFurtherHamsterAt(mpw::Location location) {
    std::make_shared<Hamster>(game->getTerritory(), location, mpw::Direction::EAST, 0);
}

void HamsterViewTestBase::assertLayersAt(mpw::Location location, const std::vector<std::string>& expectedLayerImageNames) {
    std::string expected = "";
    for (auto& expectedName : expectedLayerImageNames) {
        expected += expectedName + "\n";
    }
    std::string actual = "";
    auto cell = viewModel->getCellAt(location.getRow(), location.getColumn());
    for (auto& layer : cell->getLayers()) {
        if (layer.getVisible()) {
            actual += layer.getImageName() + "\n";
        }
    }
    EXPECT_EQ(expected, actual);
}

void HamsterViewTestBase::assertButtons(const std::string& expected) {
    std::string actual = HamsterViewModelStringifier::buttonBarToExpectationString(*viewModel);
    EXPECT_EQ(expected, actual);
    assertLocationsAreSet();
    assertSizeIsConsistent();
}
